text/0000-undo-universal-impl-trait.md

```diff
@@ -0,0 +1,238 @@
+- Feature Name: undo-universal-impl-trait
+- Start Date: 2018-05-21
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+This RFC undoes [RFC 1951], which gave Rust the `impl Trait` syntax in function argument position.
+
+# Motivation
+[motivation]: #motivation
+
+## Brief history
+
+`impl Trait` was initially introduced in [RFC 1522]. That feature was completely new to Rust as it
+added a very interesting and powerful feature: being able to provide the caller of a function with a
+type contract by providing *an anonymized type* in return location of the function and by letting
+the callee pick the type.
+
+    // the caller only knows _properties_ on the returned value, it doesn’t know its real type
+    fn zero_to_ten() -> impl Iterator<Item = u32>
+
+That feature was a complete success and was quickly adopted. The syntax, `impl Trait`, was then
+quickly refered as “the syntax to pick the type in the callee” and was very simple to get our
+fingers around. This is a use of *existential quantification* and the single one currently available
+in Rust. With the right education – and the Rust team made a wonderful job with the book – it’s easy
+to teach newcomers about the feature.
+
+Then came [RFC 1951] and [RFC 2071]. The latter generalizes the use of `impl Trait` to `let`
+bindings, `const` and `static` declarations and is not of our matter here – because they also act as
+*existentials*, which is great. However, the former, [RFC 1951], introduced `impl Trait` in argument
+position. There were a lot of debates about the semantics that the syntax should have but
+eventually, the consensus came to a decision. The following syntax:
+
+    fn foo(x: impl Debug);
+
+Is akin to this:
+
+    fn foo<T: Debug>(x: T)
+    fn foo<T>(x: T) where T: Debug
+
+And so the feature was stabilized.
+
+This RFC wants to undo this.
+
+## Why is `impl Trait` in argument position wrong?
+
+People out there didn’t really realize what was going on. Complains and confusions on IRC, reddit
```

Comment:
This feels like a straw man to me. The feature is not yet documented. There is always going to be confusion until that gets remedied. There's also a lot of confusion around how reborrowing works (another undocumented feature), and that feature's been around for much longer.